src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslPkcs12Reader.cs

```diff
@@ -28,141 +31,77 @@ private OpenSslPkcs12Reader(SafePkcs12Handle pkcs12Handle)
+                        privateKeyInfo.PrivateKeyAlgorithm.Algorithm.Value);
             }
 
-            if (_evpPkeyHandle != null)
-            {
-                _evpPkeyHandle.Dispose();
-                _evpPkeyHandle = null;
-            }
+            key.ImportPkcs8PrivateKey(pkcs8.Span, out int bytesRead);
 
-            if (_pkcs12Handle != null)
+            if (bytesRead != pkcs8.Length)
             {
-                _pkcs12Handle.Dispose();
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
             }
-        }
 
-        public void Decrypt(SafePasswordHandle password)
-        {
-            bool parsed = Interop.Crypto.Pkcs12Parse(
-                _pkcs12Handle,
-                password,
-                out _evpPkeyHandle,
-                out _x509Handle,
-                out _caStackHandle);
-
-            if (!parsed)
-            {
-                throw Interop.Crypto.CreateOpenSslCryptographicException();
-            }
+            return key;
         }
 
-        public List<OpenSslX509CertificateReader> ReadCertificates()
+        internal static SafeEvpPKeyHandle GetPrivateKey(AsymmetricAlgorithm key)
         {
-            var certs = new List<OpenSslX509CertificateReader>();
-
-            if (_caStackHandle != null && !_caStackHandle.IsInvalid)
+            if (key is RSAOpenSsl rsa)
             {
-                int caCertCount = Interop.Crypto.GetX509StackFieldCount(_caStackHandle);
-
-                for (int i = 0; i < caCertCount; i++)
-                {
-                    IntPtr certPtr = Interop.Crypto.GetX509StackField(_caStackHandle, i);
-
-                    if (certPtr != IntPtr.Zero)
-                    {
-                        // The STACK_OF(X509) still needs to be cleaned up, so upref the handle out of it.
-                        certs.Add(new OpenSslX509CertificateReader(Interop.Crypto.X509UpRef(certPtr)));
-                    }
-                }
+                return rsa.DuplicateKeyHandle();
             }
 
-            if (_x509Handle != null && !_x509Handle.IsInvalid)
+            if (key is DSAOpenSsl dsa)
             {
-                // The certificate and (if applicable) private key handles will be given over
-                // to the OpenSslX509CertificateReader, and the fields here are thus nulled out to
-                // prevent double-Dispose.
-                OpenSslX509CertificateReader reader = new OpenSslX509CertificateReader(_x509Handle);
-                _x509Handle = null;
-
-                if (_evpPkeyHandle != null && !_evpPkeyHandle.IsInvalid)
-                {
-                    reader.SetPrivateKey(_evpPkeyHandle);
-                    _evpPkeyHandle = null;
-                }
-
-                certs.Add(reader);
+                return dsa.DuplicateKeyHandle();
             }
 
-            return certs;
+            return ((ECDiffieHellmanOpenSsl)key).DuplicateKeyHandle();
```

Comment:
Nit: for what it's worth, you can simplify these kinds of patterns with switch expressions:
```C#
internal static SafeEvpPKeyHandle GetPrivateKey(AsymmetricAlgorithm key) =>
    key switch
    {
        RSAOpenSsl rsa => rsa.DuplicateKeyHandle(),
        DSAOpenSsl dsa => dsa.DuplicateKeyHandle(),
        _ => ((ECDiffieHellmanOpenSsl)key).DuplicateKeyHandle()
    };
```